crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -139,35 +144,58 @@ impl SystemStage {
         self
     }
 
-    pub fn with_system_set(mut self, system_set: SystemSet) -> Self {
-        self.add_system_set(system_set);
-        self
-    }
-
-    pub fn with_run_criteria<S: System<In = (), Out = ShouldRun>>(mut self, system: S) -> Self {
-        self.system_sets[0].run_criteria.set(Box::new(system));
+    pub fn add_system(&mut self, system: impl Into<SystemDescriptor>) -> &mut Self {
+        self.add_system_inner(system, None);
         self
     }
 
-    pub fn add_system_set(&mut self, system_set: SystemSet) -> &mut Self {
+    fn add_system_inner(
+        &mut self,
+        system: impl Into<SystemDescriptor>,
+        set_run_criteria_label: Option<BoxedRunCriteriaLabel>,
+    ) {
         self.systems_modified = true;
-        let SystemSet {
-            run_criteria,
-            mut descriptors,
-        } = system_set;
-        let set = self.system_sets.len();
-        self.system_sets.push(VirtualSystemSet {
-            run_criteria,
-            should_run: ShouldRun::No,
-        });
-        for system in descriptors.drain(..) {
-            self.add_system_to_set(system, set);
+        match system.into() {
+            SystemDescriptor::Exclusive(mut descriptor) => {
+                let insertion_point = descriptor.insertion_point;
+                let run_criteria_label = descriptor
+                    .run_criteria
+                    .take()
+                    .map(|criteria| self.process_run_criteria(criteria))
```

Comment:
> As a user, I would expect adding criteria to a system inside one of these sets to apply that criteria "in addition to" the system set criteria (ex: every_other_time criteria on a system should run every other time a fixed timestep occurs).

I definitely agree here.

> After this pr I would like us to explore the idea of multiple run criteria (ShouldRun::Yes + ShouldRun::YesAndLoop == ShouldRunYesAndLoop, ShouldRun::Yes + ShouldRun::No == ShouldRun::No, etc), which would hopefully solve this problem.

@TheRawMeatball, @Ratysz and I talked about this at length during #1144 on Discord. Our conclusion was that no obvious and clearly correct logic table exists for this quaternary logic :(

Reply:
The rule encoded here is "systems' own descriptors are applied on top of the descriptor of their set" - which means overwriting any non-additive field; run criteria is just the only one such field we have right now. `SystemSet`s are not intended to be any sort of a hierarchy block or a container: they are now used *only* for mass description. Could probably do with a rename to avoid this confusion.

Reply:
I totally understand the rule (i fully groked the "system sets are weapons of mass system description). Thats not why I left the comment. The behavior in the pr is "consistent", I just think it produces surprising behaviors. I think the FixedUpdate and State lifecycle criterias will be the two most common cases for SystemSets. People using sets under those circumstances are using them to group systems that should run under a certain condition. With that mindset, adding another criteria inside the set will likely be motivated by needing "additional" criteria (otherwise the user wouldn't be putting the system in a "fixed update" system set). Replacing the criteria defeats the purpose the set was created for.

I would rather fail in those cases to tell users they did something wrong than produce surprising results. I would also prefer to not fail (but I understand that this particular case is challenging so I'm cool with tabling the multiple run criteria conversation).

Renaming won't do anything to solve the issue above, or the fact that labels do work as "expected"  by "adding" but run criteria "replace".

Reply:
My previous comment had one more paragraph that got cut off 🤦  Network's been crapping out all morning. The gist was that I share your concerns but I don't have a good way of addressing them yet, hence this behavior - there can't be just *no* behavior. Attempt to reproduce it:

I had thought about forcing the criteria defined on systems that are inside sets to be "pipable", but that seemed more arbitrary than overwriting. It can't be strait-up forbidden because having to move a system outside of a set just to be able to set its run criteria is unergonomic - you'd end up with another copy of the descriptor, all because of the very thing that was supposed to prevent having to copy the descriptor. It also can't be "do it, but warn in console": every time we do that we have to think about how to silence the warning for legit cases.

In the short-term, until we figure it out (probably via a dedicated system set API improvement PR - system sets v3?..), I'm more okay with this overwriting behavior than I would be with forbidding this outright.

Reply:
> I had thought about forcing the criteria defined on systems that are inside sets to be "pipable", but that seemed more arbitrary than overwriting.

Yeah I agree. Seems too complicated / arbitrary.

> It also can't be "do it, but warn in console": every time we do that we have to think about how to silence the warning for legit cases.

Agreed

> It can't be strait-up forbidden because having to move a system outside of a set just to be able to set its run criteria is unergonomic - you'd end up with another copy of the descriptor, all because of the very thing that was supposed to prevent having to copy the descriptor.

I don't think this transform is particularly unergonomic. And it significantly increases clarity:

```rust
// before
app
    .add_system_set(SystemSet::parallel()
        .with_run_criteria(FixedTimestep::step(0.4)
        .with_system(a.system())
        .with_system(b.system())
        .with_system(c.system().with_run_criteria(State::on_enter())
    )

// after
app
    .add_system_set(SystemSet::parallel()
        .with_run_criteria(FixedTimestep::step(0.4)
        .with_system(a.system())
        .with_system(b.system())
    )
    .add_system(c.system().with_run_criteria(State::on_enter()))
```

Reply:
For a case where you only have run criteria, sure. But consider this:
```rust
// before
app
    .add_system_set(SystemSet::parallel()
        .with_run_criteria(FixedTimestep::step(0.4))
        .label(MyLabel::Logic)
        .after(MyLabel::Physics)
        .before(MyLabel::Rendering)
        .with_system(a.system())
        .with_system(b.system())
        .with_system(c.system().with_run_criteria(State::on_enter(Foo::Bar))
    )

// after
app
    .add_system_set(SystemSet::parallel()
        .with_run_criteria(FixedTimestep::step(0.4))
        .label(MyLabel::Logic)
        .after(MyLabel::Physics)
        .before(MyLabel::Rendering)
        .with_system(a.system())
        .with_system(b.system())
    )
    .add_system(c.system()
        .with_run_criteria(State::on_enter(Foo::Bar))
        .label(MyLabel::Logic)
        .after(MyLabel::Physics)
        .before(MyLabel::Rendering)
    )
```

Reply:
Yeah thats a fair point. I'd argue that `c.system()` is in a "similar" set, but not an identical set, due to the differing run criteria. Regardless it sounds like we've agreed on the implementation (at least as a short term solution while we work the details of future run criteria improvements).